### ISSUE TITLE
KallistiOS' printf implementation hangs on normalized doubles

### ISSUE THREAD

**Issue by snickerbockers** [washingtondc-emu/washingtondc#7]
The following program hangs during the third invocation of printf in WashingtonDC but not on hardware.  

The functions where the hang is actually happenign are __pow5mult and __multiply in newlib.  I added a manual invocation of this with the same parameters that they were receiving from the printf calls (see commented out code), and the result was that it hangs on both the emulator and real hardware in the exact same way; from this it follows that the root-cause is in the parameters that printf is sending to these functions; evidently they are not the same in WashingtonDC as in real hardware.

```
#include <kos.h>
#include <math.h>
#include <GL/gl.h>
#include <GL/glut.h>
#include <inttypes.h>

KOS_INIT_ROMDISK(KOS_INIT_ROMDISK_NONE);

int main(int argc, char **argv) {
    /* _Bigint frak = { */
    /*     ._next = NULL, */
    /*     ._k = 1, */
    /*     ._maxwds = 2, */
    /*     ._sign = 0, */
    /*     ._wds = 1, */
    /*     ._x = {1} */
    /* }; */
    /* pow5mult(_REENT, &frak, 1292914048); */

    printf("For reference, 1.337 is %f\n", 1.337);
    printf("For reference, 1.0 is %f\n", 1.0);
    printf("for reference, 0.9 is %f\n", 0.9);
    printf("for reference, 0.1 is %f\n", 0.1);
    printf("for reference, 0.01 is %f\n", 0.01);
    printf("For reference, 0.001 is %f\n", 0.001);

    printf("done printing\n");

    return 0;
}
```

This is the output on WashingtonDC (it hangs printing 0.9):

```

--
KallistiOS Git revision 374a47b: Sat Jun 24 18:39:27 PDT 2017
  jay@sbckrs_desktop:/home/jay/src/KallistiOS
maple: active drivers:
    Dreameye (Camera): Camera
    Sound Input Peripheral: Microphone
    PuruPuru (Vibration) Pack: JumpPack
    VMU Driver: Clock, LCD, MemoryCard
    Mouse Driver: Mouse
    Keyboard Driver: Keyboard
    Controller Driver: Controller
    Lightgun: LightGun
  DMA Buffer at ac0c0be0
vid_set_mode: 640x480IL NTSC
maple: attached devices:
For reference, 1.337 is 1.337000
For reference, 1.0 is 1.000000
```

This is the output on a real Dreamcast:
```

--
KallistiOS Git revision 374a47b: Sat Jun 24 18:39:27 PDT 2017
  jay@sbckrs_desktop:/home/jay/src/KallistiOS
maple: active drivers:
    Dreameye (Camera): Camera
    Sound Input Peripheral: Microphone
    PuruPuru (Vibration) Pack: JumpPack
    VMU Driver: Clock, LCD, MemoryCard
    Mouse Driver: Mouse
    Keyboard Driver: Keyboard
    Controller Driver: Controller
    Lightgun: LightGun
  DMA Buffer at ac0c0be0
vid_set_mode: 640x480IL NTSC
dc-load console support enabled
maple: attached devices:
  A0: Dreamcast Controller          (01000000: Controller)
  A1: Visual Memory                 (0e000000: Clock, LCD, MemoryCard)
  A2: Puru Puru Pack                (00010000: JumpPack)
  B0: Dreamcast Controller          (01000000: Controller)
  B1: Visual Memory                 (0e000000: Clock, LCD, MemoryCard)
For reference, 1.337 is 1.337000
For reference, 1.0 is 1.000000
for reference, 0.9 is 0.900000
for reference, 0.1 is 0.100000
for reference, 0.01 is 0.010000
For reference, 0.001 is 0.001000
done printing
arch: shutting down kernel
maple: final stats -- device count = 5, vbl_cntr = 4, dma_cntr = 4
vid_set_mode: 640x480IL NTSC
```

**Comment by snickerbockers** [comment]
The change in behavior between WashingtonDC and Dreamcast appears to originate in this line of code in dtoa.c, which resides within newlib within KallistiOS (for reference, _DOUBLE_IS_32BITS is actually defined here; not sure if that actually means that double is 32-bits or not):

```
#if defined (_DOUBLE_IS_32BITS)
  ds = (d2.d - 1.5) * 0.289529651 + 0.176091269 + i * 0.30103001;
#else
  ds = (d2.d - 1.5) * 0.289529654602168 + 0.1760912590558 + i * 0.301029995663981;
#endif                                                                                                                                                                                                                                        
  k = (int) ds;                                                                                                                                                                                                                               
```

d2.d is the input value supplied to dtoa.  i is 0xffffffff.  The final value of k is 0 on real hardware and 0x4d104d80 on WashingtonDC

**Comment by snickerbockers** [comment]
It's the "float fpul, frn" opcode

```
(gdb) print $fpul                                                                                                                                                                                                                             
$3 = -nan(0x7fffff)                                                                                                                                                                                                                           
(gdb) print $fr1                                                                                                                                                                                                                              
$4 = 4.2949673e+09                                                                                                                                                                                                                            
(gdb) print (float)0x7fffff                                                                                                                                                                                                                   
$5 = 8388607                                                                                                                                                                                                                                  
(gdb) x/i $pc                                                                                                                                                                                                                                 
   0x8c01059e:  fmul    fr2,fr1                                                                                                                                                                                                               
(gdb) x/i $pc-2                                                                                                                                                                                                                               
   0x8c01059c:  float   fpul,fr1                                                                                                                                                                                                              
```

**Comment by snickerbockers** [comment]
Fixed by b2e529d234052512bb4b290262c247fde1f7e84e
